mpcs/src/basefold/commit_phase.rs

```diff
@@ -86,97 +90,108 @@ where
-            let running_root = commitment;
-            write_digest_to_transcript(&running_root, transcript);
-            roots.push(running_root);
-
-            // go next round sumcheck and update sumcheck message
-            running_sumcheck_message =
-                sum_check_challenge_round(&mut eq, &mut running_evals, challenge);
-        } else {
-            // Assign a new value to the old running vars so that the compiler
-            // knows the old value is safe to move.
-            running_sumcheck_message = Vec::new();
-            // The difference of the last round is that we don't need to compute the message,
-            // and we don't interpolate the small polynomials. So after the last round,
-            // running_evals is exactly the evaluation representation of the
-            // folded polynomial so far.
-            sum_check_last_round(&mut eq, &mut running_evals, challenge);
-            transcript.append_field_element_exts(&running_evals);
-            final_evals = running_evals;
-            // To avoid the compiler complaining that running_evals is moved.
-            running_evals = Vec::new();
-
-            if cfg!(feature = "sanity-check") {
-                // If the prover is honest, in the last round, the running oracle
-                // on the prover side should be exactly the encoding of the folded polynomial.
-                let evaluations = final_evals.clone();
-                let basecode = <Spec::EncodingScheme as EncodingScheme<E>>::encode_slow_ext(
-                    p3::matrix::dense::DenseMatrix::new(evaluations, 1),
-                );
-                assert_eq!(basecode.values, new_running_oracle.values);
-            }
-        }
-        end_timer!(sumcheck_timer);
+
+    let sumcheck_phase1 = entered_span!("sumcheck_phase1");
+    for i in 0..num_rounds.min(num_vars - log2_strict_usize(num_threads)) {
+        challenge = basefold_one_round::<E, Spec>(
+            pp,
+            &mut prover_states,
+            challenge,
+            &mut sumcheck_messages,
+            &initial_oracle,
+            transcript,
+            &mut trees,
+            &mut roots,
+            &mmcs_ext,
+            i == num_rounds - 1,
+        );
+    }
+    exit_span!(sumcheck_phase1);
+
+    if let Some(p) = challenge {
+        prover_states
+            .iter_mut()
+            .for_each(|prover_state| prover_state.fix_var(p.elements));
+    }
+
+    // deal with log(#thread) basefold rounds
+    let merge_sumcheck_polys_span = entered_span!("merge_sumcheck_polys");
+    let poly = merge_sumcheck_polys(&prover_states);
+    let mut prover_states = vec![IOPProverState::prover_init_with_extrapolation_aux(
+        poly,
+        vec![(vec![], vec![])],
+    )];
+    exit_span!(merge_sumcheck_polys_span);
+
+    let mut challenge = None;
+
+    let sumcheck_phase2 = entered_span!("sumcheck_phase2");
+    for i in 0..num_rounds.saturating_sub(num_vars - log2_strict_usize(num_threads)) {
```

Comment:
nitpick: allocate a new variable like `remaining_rounds` to store `num_rounds.saturating_sub(num_vars - log2_strict_usize(num_threads))`.